keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2469 @@
+  but supporting sharing between pods would need a completely new
+  API similar to the one in this KEP.
+- Using a resource that is expensive to initialize multiple times
+  in different pods. This is not possible at the moment.
+- Custom parameters that describe resource requirements and initialization.
+  Parameters are not limited to a single, linear quantity that can be counted.
+  With the current Pod API, annotations have to be used to capture such
+  parameters and then hacks are needed to access them from a CSI driver or
+  device plugin.
+
+Support for new hardware will be provided by hardware vendor add-ons. It will
+not be necessary anymore to modify Kubernetes itself.
+
+This KEP does not replace other means of requesting traditional resources
+(RAM/CPU, volumes, extended resources). The scheduler will serve as coordinator
+between the add-ons which own resources (CSI driver, resource driver) and the
+resources owned and assigned by the scheduler (RAM/CPU, extended resources).
+
+## Motivation
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The kubelet device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources is too limited. This KEP addresses
+limitations of the current approach for the following use cases:
+
+- *Device initialization*: When starting a workload that uses
+  an accelerator like an FPGA, I’d like to have the accelerator
+  reconfigured or reprogrammed for the workload before the workload
+  itself starts. For security reasons, workloads should not be able to
+  reconfigure devices directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+  For the FPGA example, a file containing the desired configuration
+  of the FPGA has to be referenced.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown. For example, an FPGA might have
+  to be reset because its configuration for the workload was
+  confidential.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
+  to use part of the shareable device inside a container and other
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow one to allocate part of the device because parameters are too limited
+  and Kubernetes doesn't have enough information about the extended
+  resources on a node to decide whether they can be shared.
```

Comment:
```suggestion
*Limitation*: For example, newer generations of NVIDIA GPUs have a mode of
operation called MIG, that allow them to be sub-divided into a set of mini-GPUs
(called MIG devices) with varying amounts of memory and compute resources
provided by each. From a hardware-standpoint, configuring a GPU into a set of
MIG devices is highly-dynamic and creating a MIG device tailored to the
resource needs of a particular application is well supported. However, with the
current device plugin API, the only way to make use of this feature is to
pre-partition a GPU into a set of MIG devices and advertise them to the kubelet
in the same way a full / static GPU is advertised. The user must then pick from
this set of pre-partitioned MIG devices instead of having one created for them
on the fly based on their particular resource constraints. Without the ability
to create MIG devices dynamically (i.e. at the time they are requested) the set
of pre-defined MIG devices must be carefully tuned to ensure that GPU resources
do not go unused because some of the pre-partioned devices are in low-demand.
It also puts the burden on the user to pick a particular MIG device type,
rather than declaring the resource constraints more abstractly.
```

Reply:
Thanks, looks good to me.

I have added this as a separate commit to a pending update for this KEP, see https://github.com/pohly/enhancements/pull/16